WireUI/Sources/WireUIFoundation/Main/MainCoordinator.swift

```diff
@@ -0,0 +1,364 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import UIKit
+import WireFoundation
+
+// swiftlint:disable opening_brace
+
+/// Manages the main navigation and the layout changes of the application after a successful login.
+///
+/// The MainCoordinator class is the central controller for the app's navigation and layout management.
+/// It receives references to ``MainTabBarControllerProtocol`` and ``MainSplitViewControllerProtocol``
+/// conforming instances and is responsible for managing transitions between different split layout states (collapsed and expanded)
+/// as well as handling navigation logic.
+///
+/// Both, the split view controller as well as the tab view controller actually install `UINavigationController`
+/// instances and then put or remove the content view controllers into/from `viewControllers` array.
+
+@MainActor
+public final class MainCoordinator<
+
+    SplitViewController: MainSplitViewControllerProtocol,
+    TabBarController: MainTabBarControllerProtocol,
+    NewConversationBuilder: MainCoordinatorInjectingViewControllerBuilder,
+    SelfProfileBuilder: MainCoordinatorInjectingViewControllerBuilder
```

Comment:
do we need to have two dedicated builders since they're conforming to the same protocol, we could have just one it would make the method signature easier to understand since the protocol is basically a view controller build builder.

```
ViewControllerBuilder: MainCoordinatorInjectingViewControllerBuilder 
```

Reply:
Yes exactly,  the single builder would be responsible of building the various controllers related to the main coordinator, you would still be able to inject it, the MainCoordinator signature would be a little simpler while still providing a well-defined scoped of what a main coordinator builder should do.

But if you planned it differently by having separated builders, it's fine too 👍.